[feature] perform captive portal authentication by submitting a form

Vivekrajput20 opened the pull request:
Closes #60, closes #61, closes #66, closes #67 , closes #68 
Depends on [openwisp-radius#130](https://github.com/openwisp/openwisp-radius/pull/130)

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -9,47 +9,123 @@ import { Link } from 'react-router-dom';
+    }
+  }
+
+
```
looks like there's one extra blank line

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -108,6 +184,23 @@ export default class Status extends React.Component {
         </div>
+        {captivePortalForm && (window.top === window.self) &&
+          <>
+            <form ref={this.formRef} method={captivePortalForm.method || ''}
```
why does this default to empty string?

nemesisdesign commented on internals/generators/organization/config.yml.hbs:
```diff
@@ -323,6 +324,20 @@ client:
+        - name: zone
+          value: zone_name
+        - name: redirurl
+          value: http://0.0.0.0:8080/{{slug}}/status
```
isn't `localhost` or `127.0.0.1` more common? It may be less confusing for most users.

Vivekrajput20 replied:
Currently it's passed as an array field and it's name can be different for different captive portals, so I don't have any way to identify this field and modify it's value on fronted app. We can ask for captive portal's root url and save it in yml file and then use that variable in all the places where relative url is required.

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -108,6 +184,23 @@ export default class Status extends React.Component {
+                <input readOnly type="text" name={field.name} value={field.value} />
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title='owisp-auth-iframe' />
```
you're using single quotes for `owisp-auth-iframe` while double quotes for all other attrs

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
+                <input readOnly type="text" name={field.name} value={field.value} />
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title="owisp-auth-iframe" />
```
can we indent this a bit more to avoid the line being too long?

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
           </div>
         </div>
+        {captivePortalForm && (window.top === window.self) &&
+          <>
```
what does this mean? I should refresh my react fu, maybe you can help me?

Vivekrajput20 replied:
It is possible that captive portal will redirect to the status page, in that case status page will go into a infinite loop where each IFrame will render a status page + a new Iframe, to break that I've added this check to execute this block only in root window.

nemesisdesign replied:
can we add a comment?

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title="owisp-auth-iframe" />
+          </>
```
also here you can help me out refreshing my react fu, why is this needed?

Vivekrajput20 replied:
ref is passed to get a reference to that element inside react class, I'm submitting the form in this iframe and have added this onLoad handler which checks for  reply param and 404 after the response is received from captive protal and loaded in the Iframe.

nemesisdesign replied:
also here can we add a comment?

Vivekrajput20 commented on internals/generators/organization/config.yml.hbs:
```diff
@@ -323,6 +324,20 @@ client:
 
+    captive_portal_form:
+      method: post
+      action: http://localhost/
```
Sure

Vivekrajput20 commented:
@nemesisdesign  please test the lastest PR, I've updated login with saving macaddr in session and also added logout functionality.

Vivekrajput20 commented on client/components/status/status.js:
```diff
@@ -1,55 +1,178 @@
+    const macaddr = cookies.get(`${orgSlug}_macaddr`);
+    try {
+      const response = await axios({
+        method: "post",
```
@nemesisdesign  this is a request to proxy server and proxy server actually sends a GET request only, I'll convert it to GET request.

Vivekrajput20 replied:
This was a request to proxy server and proxy server was sending the GET request only. I've updated it to GET in the latest push.

Vivekrajput20 commented:
@nemesisdesign  It works in my local so I can't exactly figure out what's the bug, the configuration is fine and should have worked.
These are some of the things that can be causing this error.
-  I added `@babel/plugin-transform-runtime` package in this PR, so this PR needs npm install before proceeding.
- Frontend server starts at `0.0.0.0:8080/` and if the captive portal redirurl is `http://localhost:8080/default/status` and not `http://0.0.0.0:8080/default/status` this won't work as a document can access an Iframe's url only if domain is same. And 0.0.0.0 and localhost are considered different domains.

I've improved the PR with few more checks.

Vivekrajput20 commented:
@nemesisdesign that localhost is defined in default org's configuration ([here](https://github.com/openwisp/openwisp-wifi-login-pages/blob/4d1fa882cd3232a5ab59bc163f29703a48cc3054/org-configurations/default-configuration.yml#L301))
I think you are accessing captive portal for default organization that's why this request to local host is made. You will have to change the organization slug in the browser url.

Vivekrajput20 commented:
@nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
I added a loader as well after the logout button is clicked.
Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

nemesisdesign commented:
> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

Ah ok I understand, there was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

> Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

Ok thanks.

nemesisdesign reviewed (commented):
> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

Ah ok I understand, there was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

> Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

Ok thanks.

nemesisdesign reviewed (changes requested):
Thanks for the update @Vivekrajput20,

I tried running the latest version of this branch but I get this error when I run `npm start`:

```
npm ERR! code ELIFECYCLE
npm ERR! errno 1
[0] npm ERR! openwisp-wifi-login-pages@1.0.0 server: `npm run setup && nodemon server/start.js`
[0] npm ERR! Exit status 1
npm ERR! 
[0] npm ERR! Failed at the openwisp-wifi-login-pages@1.0.0 server script.
[0] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[0] 
[0] npm ERR! A complete log of this run can be found in:
[0] npm ERR!     /home/nemesis/.npm/_logs/2020-09-06T22_37_36_382Z-debug.log
[0] npm run server exited with code 1
[1] No server url specified. Expecting server to run on http://localhost:3030
[1] ℹ ｢wds｣: Project is running at http://0.0.0.0:8080/
[1] ℹ ｢wds｣: webpack output is served from /
[1] ℹ ｢wds｣: Content not from webpack is served from /home/nemesis/Code/openwisp/openwisp-wifi-login-pages/public
[1] ℹ ｢wds｣: 404s will fallback to /index.html
[1] events.js:174
[1]       throw er; // Unhandled 'error' event
[1]       ^
[1] 
[1] Error: ENOSPC: System limit for number of file watchers reached, watch '/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/public'
[1]     at FSWatcher.start (internal/fs/watchers.js:165:26)
[1]     at Object.watch (fs.js:1258:11)
[1]     at createFsWatchInstance (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:38:15)
[1]     at setFsWatchListener (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[1]     at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:233:14)
[1]     at FSWatcher.NodeFsHandler._handleDir (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:429:19)
[1]     at FSWatcher.<anonymous> (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:477:19)
[1]     at FSWatcher.<anonymous> (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:482:16)
[1]     at FSReqWrap.oncomplete (fs.js:154:5)
[1] Emitted 'error' event at:
[1]     at FSWatcher._handleError (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/index.js:260:10)
[1]     at createFsWatchInstance (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:40:5)
[1]     at setFsWatchListener (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[1]     [... lines matching original stack trace ...]
[1]     at FSReqWrap.oncomplete (fs.js:154:5)
[1] npm ERR! code ELIFECYCLE
[1] npm ERR! errno 1
[1] npm ERR! openwisp-wifi-login-pages@1.0.0 client: `npm run setup && webpack-dev-server --host 0.0.0.0 --mode development --config config/webpack.config.js`
[1] npm ERR! Exit status 1
```
Do you know what is going on? How can I fix this?

Also, I want to underline what I wrote in the previous comment:

> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

There was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

### Tests

I think the solution is very near completion, the login works, the logout should work after your changes are fixed, the `macaddr` param issue should be easy to fix, I think this is a good time to fix the tests as well.

Vivekrajput20 commented:
@nemesisdesign  the fix for this issue is mentioned [here](https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756).
I'll add support for query sting in the status url. And will also start fixing the tests.

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -1,55 +1,178 @@
+        const macaddr = cookies.get(`${orgSlug}_macaddr`);
+        if (macaddr) {
+          await this.getUserRadiusSessions();
+          if (sessions && sessions.length === 0) {
```
I think we should add a comment above this to explain we'll do the log in only if there's no other active session.
BTW I think we should use the mac address param (if not empty) here to filter the sessions, because otherwise users may not be able to log in with two different devices using the same account

Vivekrajput20 replied:
@nemesisdesign  Since mac address is stored in a cookie, it's already available to the proxy server and there was no need to send it from front-end. I've already added this filter to [proxy server's code](https://github.com/openwisp/openwisp-wifi-login-pages/blob/432e88b4c7913b129f5b30835535c965a86c3d47/server/controllers/user-radius-sessions-controller.js#L23)
And I'll add the comment explaining that.

nemesisdesign reviewed (commented):
> @nemesisdesign the fix for this issue is mentioned [here](https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756).

Thanks, that worked.

I tried again but the POST request to the captive portal logout is not firing, even though the code looks ok:

![logout-broken](https://user-images.githubusercontent.com/841044/92433066-9f99cd00-f161-11ea-9f6e-01822219292a.gif)

> I'll add support for query string in the status url. And will also start fixing the tests.

Ok, thanks!

Vivekrajput20 commented:
> I tried again but the POST request to the captive portal logout is not firing, even though the code looks ok
@nemesisdesign It must be firing logout request but as you have selected only XHR requests in the network that is why it's not showing. Since it's a HTML request it will be visible in All requests.

nemesisdesign commented on client/components/status/status.js:
```diff
@@ -1,55 +1,191 @@
+    }
+    logout(cookies, orgSlug);
+    setLoading(false);
+    toast.success(logoutSuccess);
```
if I comment out these 3 lines (106, 107 and 108), the captive portal logout works.
I think these lines should be executed only once `this.logoutFormRef.current.submit()` has finished, otherwise it seems the submit is not triggered.
Can you add listener to the onload of the iframe or anything similar so we can execute these 3 lines only after the captive portal logout has completed?

Keep in mind the same should happen for the login as well, if the captive portal login has to be executed, the success message should be shown only once the captive portal has replied, otherwise users may try to visit another page and be redirected again to the captive page and then they'll be super confused.

With this solved we're ready to merge and move on! :+1: :clap:

nemesisdesign reviewed (changes requested):
Great progress Vivek, thank you!

I tested and debugged it again and found the reason for which the captive portal logout is not working, I left a comment below.

I also spotted this console warning:

![Screenshot from 2020-09-09 21-30-11](https://user-images.githubusercontent.com/841044/92675361-d0eed600-f2e4-11ea-94dd-ff3c2e9503ab.png)

nemesisdesign reviewed (approved):
It works! Thanks @Vivekrajput20 :+1: :clap: :tada:
